bpo-26510: Allow argparse add_subparsers to take `required` kwarg.

asottile opened the pull request:
I added both a test to demonstrate the current behaviour / workaround as well as the new behaviour.

I'm torn between setting the default `required=True` (which was the python2 behaviour) and setting `required=False` (the current python3 behaviour).

If we're ok with `required=True`, I'll change that up :)

<!-- issue-number: bpo-26510 -->
https://bugs.python.org/issue26510
<!-- /issue-number -->

asottile committed:
Allow argparse add_subparsers to take `required` kwarg.

asottile commented:
Actually on second thought, I'm going to try and restore the python 2 behaviour -- optional subparsers *rarely* make sense (and it'll will make writing py2 + py3 simpler).

asottile committed:
Subparsers are required by default.

This fixes a regression introduced between python2 and python3.

tseaver reviewed (approved):
Agreed that `required=True` is the better default.

merwok commented on Lib/argparse.py:
```diff
@@ -1064,6 +1064,7 @@ def __init__(self,
                  dest=SUPPRESS,
+                 required=True,
                  help=None,
                  metavar=None):
```
For maximum backward compatibility, the new parameter should be added to the end.

(I don’t personnally write calls like `something('abc', SUPPRESS, True, None, None)` but they are supported, and should not be broken when possible.)

asottile replied:
`add_subparsers` itself takes no keyword arguments so I don't think there's a concern of the private api here.

```python
>>> parser.add_subparsers('foo', True, False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: add_subparsers() takes 1 positional argument but 4 were given
```

(the 1 positional argument here being `self`)

The ordering I chose was intentional to match the signatures in the rest of the file.

merwok replied:
Perfect!

asottile replied:
@merwok do I monty-python it up now?  I'm unclear on the process for dismissing a review :P

merwok replied:
The review holding this up is «please update docs and news» :)

asottile replied:
whoops totally missed that! I'll get right on it :)

merwok reviewed (changes requested):
Please update docs and news.

asottile committed:
Add news and update docs.

asottile commented:
[I didn't expect the Spanish Inquisition!](https://youtu.be/hNavxvvN9QI)

merwok committed:
Minor tweaks

merwok commented:
This looks all right!  Just to be exhaustive, are the interactions ok with required options and required subparses?  (e.g. error messages mention both the missing required option and the missing required subparser)

merwok commented on Misc/NEWS.d/next/Library/2017-09-19-13-29-29.bpo-26510.oncW6V.rst:
```diff
@@ -0,0 +1,3 @@
+Fix regression in ``add_subparsers``: subparsers are now required by
+default. For optional subparsers, ``required=False`` may be passed as a
+keyword argument to ``add_subparsers``.  Patch by Anthony Sottile
```
@Mariatta Could you tell me if it’s the current practice to include contributors’ names in NEWS entries?

@asottile Add yourself to ACKS if you’re not already there.

Mariatta replied:
@merwok Yes to `Patch by <contributor>` in news entries.

And also +1 to adding @asottile to Misc/ACKS.

Mariatta replied:
`Patch by <contributor>` should end with a period. :)

asottile replied:
I attempted to follow [this example](https://docs.python.org/devguide/committing.html#what-s-new-and-news-entries) --  if not perhaps that should be updated.

[Found myself!](https://github.com/python/cpython/blob/525269430a3f9fbb7287e4bb6b365ac216004980/Misc/ACKS#L1489)

merwok replied:
Thanks for the help Mariatta and for the link Anthony!

merwok committed:
rewrite news

asottile commented:
> This looks all right! Just to be exhaustive, are the interactions ok with required options and required subparses? (e.g. error messages mention both the missing required option and the missing required subparser)

I can check these, though they'd already be broken (not a regression) -- I'm not actually introducing any new functionality, just a new default.  let me go through the scenarios though!

asottile commented:
I fixed the only crash I could find related to this: https://github.com/python/cpython/pull/3027/commits/6e61d8a99c219ca09e070a97822201c7de814632

It was already triggerable with the following:

```python
import argparse

parser = argparse.ArgumentParser()
parsers = parser.add_subparsers()
parsers.required = True
parsers.add_parser('foo')
parsers.add_parser('bar')

parser.parse_args()
```

```
$ python3.6 test.py
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    parser.parse_args()
  File "/usr/lib/python3.6/argparse.py", line 1739, in parse_args
    args, argv = self.parse_known_args(args, namespace)
  File "/usr/lib/python3.6/argparse.py", line 1771, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib/python3.6/argparse.py", line 2006, in _parse_known_args
    ', '.join(required_actions))
TypeError: sequence item 0: expected str instance, NoneType found
```

The regression test I added alongside failed in the same way before patching (yay TDD).

Should this get called out separately?

merwok commented:
Agreed, this looks like an unrelated problem that needs its own bug and PRs.  Good catch!

My question was more related to the interaction of a required subparser with a required option, but I guess if there is any issue it was already in 2.7 and previous versions.  I’m ready to approve this when you back out the latest commit.

merwok committed:
Make explicit that there is a new param

asottile commented:
Sweet, backed out the commit -- I'll follow up with that patch as well once this is merged!
